eth_sign toggle Update in advanced settings

NidhiKJha opened the pull request:
Currently,  eth_sign has been disabled by default, but users can enable the confirmation method in Settings > Advanced. Enabling eth_sign puts the user at risk of [phishing scams](https://medium.com/chainlight/si-vis-pacem-para-bellum-exploring-metamask-phishing-4605425d80a7) that can result in loss of funds. This PR is to add extra friction to enable eth_sign by adding Modal and text input. This PR also includes the analytics update for eth_sign

1. When user completes the flow to enable eth_sign toggle, we should fire the Settings Updated event with a enable_eth_sign boolean property with true as the value.

2. When user completes the flow to disable eth_sign toggle, we should fire the Settings Updated event with a enable_eth_sign boolean property with false as the value.

Fixes #18586 

## Screenshots/Screencaps

### Before

https://user-images.githubusercontent.com/39872794/234841744-2096c3fb-ffaa-424a-ba5f-d424fae584e0.mov


### After


https://user-images.githubusercontent.com/39872794/235162191-2ff933b7-503d-433e-b36d-1e076d405401.mov




## Manual Testing Steps

-  Go to the Settings > Advanced screen
- Toggle on the eth_sign toggle button
- Complete the steps in the modal
- At any step if you cancel the flow or click outside of the modal, it won't toggle on the modal unless you click the enable button in the last step of modal
- To turn off the eth_sign just click the toggle button

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

NidhiKJha committed:
added eth sign first step

NidhiKJha committed:
added modal

NidhiKJha committed:
added validation for form

NidhiKJha committed:
updated width with block

NidhiKJha committed:
added state trigger for toggle

NidhiKJha committed:
updated Eth sign modal text changes

NidhiKJha committed:
added eth sign toggle tex

NidhiKJha committed:
removed unnecessary code

NidhiKJha committed:
fixed form validation text

NidhiKJha committed:
updated eth toggle text

NidhiKJha committed:
added test

NidhiKJha committed:
added analytics

SaraCheikh commented:
@coreyjanssen for the last step before enabling the setting, are we okay removing "try again" from the error state copy since it will be shown while they are typing the text?

SaraCheikh commented:
This looks very clean @NidhiKJha ! small details for the UI: 

-The spacing between title and body text looks very tight. In design, the default spacing in the modal component is 24px.

 <img width="487" alt="Screenshot 2023-04-28 at 13 56 27" src="https://user-images.githubusercontent.com/98161559/235141326-0a9496cd-ca50-42eb-b54f-a7128af3bea9.png">
 
 -We are using a secondary button (outlined) instead of a primary one (filled) for " cancel",  we are also using a size lg for these buttons which looks slightly bigger. 
 
<img width="447" alt="Screenshot 2023-04-28 at 14 00 40" src="https://user-images.githubusercontent.com/98161559/235142740-ec01b983-a51d-406d-9056-18a78ab702c0.png">

- @coreyjanssen has used sentence case in design for `(Not recommended)` and `(Recommended) ` int he toggle state. 

<img width="332" alt="Screenshot 2023-04-28 at 14 08 26" src="https://user-images.githubusercontent.com/98161559/235143584-50bcb3f4-2e50-4b95-83a0-e22144327839.png">

NidhiKJha committed:
updated design changes

NidhiKJha committed:
lint fix

NidhiKJha commented:
> This looks very clean @NidhiKJha ! small details for the UI:
> 
> -The spacing between title and body text looks very tight. In design, the default spacing in the modal component is 24px.
> 
@SaraCheikh I have updated the changes. I have updated the screencast video in description and you can check the Modal [here ](https://output.circle-artifacts.com/output/job/7ed58095-4d11-4489-99bb-cedfc245e929/artifacts/0/storybook/index.html?path=/story/components-app-modals-ethsignmodal--default-story)

coreyjanssen commented:
> @coreyjanssen for the last step before enabling the setting, are we okay removing "try again" from the error state copy since it will be shown while they are typing the text?

Works for me! Remove the period as well as "Try again" so it just reads `The text is incorrect` :)

NidhiKJha committed:
updated error text

darkwing reviewed (dismissed):
I'm seeing the following error when I click to sign:

<img width="1113" alt="SCR-20230428-jnnt" src="https://user-images.githubusercontent.com/46655/235189485-d08fd8b6-aa38-4230-9b6d-29c6f7888363.png">

Nothing happens upon first click. I have to click multiple times to actually sign.

SaraCheikh commented:
> > This looks very clean @NidhiKJha ! small details for the UI:
> > -The spacing between title and body text looks very tight. In design, the default spacing in the modal component is 24px.
> 
> @SaraCheikh I have updated the changes. I have updated the screencast video in description and you can check the Modal [here ](https://output.circle-artifacts.com/output/job/7ed58095-4d11-4489-99bb-cedfc245e929/artifacts/0/storybook/index.html?path=/story/components-app-modals-ethsignmodal--default-story)

Buttons look great now @NidhiKJha! only two more details:
-Remove "try again" in the copy of the error state.
-Make spacing between title and body text 24px. 
<img width="1211" alt="Screenshot 2023-05-01 at 12 48 29" src="https://user-images.githubusercontent.com/98161559/235442932-12d1c275-65c5-4960-9d8f-ab7e177b547d.png">

brad-decker reviewed (dismissed):
So, the components/app/modals pattern and using dispatch to open modals should be deprecated and marked thusly. we do not want to extend this folder or its usages, at least the last time I checked. We had a popover component that can be used instead that doesn't require a dispatch to control showing it... unless that has changed recently? @georgewrmarshall what is the current approach recommended for modal implementation?

NidhiKJha commented:
> I'm seeing the following error when I click to sign:

Hmm! I looked into this issue. This is not related to this PR but the sign button is not working. I have created a [separate issue here](https://github.com/MetaMask/metamask-extension/issues/18942)

NidhiKJha commented:
> So, the components/app/modals pattern and using dispatch to open modals should be deprecated and marked thusly. we do not want to extend this folder or its usages, at least the last time I checked. We had a popover component that can be used instead that doesn't require a dispatch to control showing it... unless that has changed recently? @georgewrmarshall what is the current approach recommended for modal implementation?

Thanks for pointing it out Brad. I have used the Modal component since we have used this component for all the modals on the settings page and it's all handled in one file. I know that @georgewrmarshall is working on the new Modal component that will deprecate the old Modals once it's merged but I wasn't aware that we are deprecating this Modals with Popover.

NidhiKJha committed:
updated changes

NidhiKJha commented on ui/components/app/modals/eth-sign-modal/eth-sign-modal.js:
```diff
@@ -0,0 +1,196 @@
+    dispatch(
+      setDisabledRpcMethodPreference(
+        'eth_sign',
+        !disabledRpcMethodPreferences?.eth_sign,
```
Updated here https://github.com/MetaMask/metamask-extension/pull/18848/commits/00d7f01f4923628abea4c6eedbefbb9b1adad7ba

NidhiKJha commented on ui/components/app/modals/eth-sign-modal/eth-sign-modal.js:
```diff
@@ -0,0 +1,196 @@
+  const t = useI18nContext();
+  const dispatch = useDispatch();
+  const trackEvent = useContext(MetaMetricsContext);
+  const handleCancel = () => {
```
Thanks for pointing it out. Yeah, we can use hideModal directly Updated [here ](https://github.com/MetaMask/metamask-extension/pull/18848/commits/00d7f01f4923628abea4c6eedbefbb9b1adad7ba)

NidhiKJha commented:
> Buttons look great now @NidhiKJha! only two more details: -Remove "try again" in the copy of the error state. -Make spacing between Icon and title 16 px and between title and body text 24px. 

@SaraCheikh  Updated! Need the builds to pass and the changes would be updated in the storybook then :)

georgewrmarshall commented:
> So, the components/app/modals pattern and using dispatch to open modals should be deprecated and marked thusly. we do not want to extend this folder or its usages, at least the last time I checked. We had a popover component that can be used instead that doesn't require a dispatch to control showing it... unless that has changed recently? @georgewrmarshall what is the current approach recommended for modal implementation?

Hello @brad-decker @NidhiKJha, I was aware of both but never asked the question which we should use. 🤦 I was currently working on a Modal component that could be used for both so this is good information. May I suggest we add a deprecation notice to the modal.js folder so other engineers are aware? I'll create an issue https://github.com/MetaMask/metamask-extension/issues/18975

NidhiKJha commented on ui/components/app/modals/eth-sign-modal/index.scss:
```diff
@@ -0,0 +1,8 @@
+
+  &__close {
+    position: absolute;
+    right: 16px;
```
Yup, it works fine for RTL. This is the same logic we used for other modals

https://user-images.githubusercontent.com/39872794/235983169-3870956b-797c-45f5-acf1-20cb2acfff9f.mov

darkwing reviewed (approved):
This is *excellent* @NidhiKJha !  Well done!
